runtime/drivers/azure/azure.go

```diff
@@ -0,0 +1,423 @@
+		}
+
+		if respErr.RawResponse.StatusCode == http.StatusForbidden && respErr.ErrorCode == "AuthorizationPermissionMismatch" {
+			c.logger.Named("Console").Warn("Azure Blob Storage account does not have permission to list blobs. Falling back to anonymous access.")
+			client, err = c.createAnonymousClient(ctx, conf)
+			if err != nil {
+				return nil, err
+			}
+
+			bucketObj, err = azureblob.OpenBucket(ctx, client, nil)
+			if err != nil {
+				return nil, err
+			}
+
+			iter, err = rillblob.NewIterator(ctx, bucketObj, opts, c.logger)
+		}
+
+		// check again
+		if errors.As(err, &respErr) && respErr.StatusCode == http.StatusForbidden {
+			return nil, drivers.NewPermissionDeniedError(fmt.Sprintf("failed to create iterator: %v", respErr))
+		}
+	}
+
+	return iter, nil
+}
+
+type sourceProperties struct {
+	Path                  string         `mapstructure:"path"`
+	AccountName           string         `mapstructure:"account_name"`
+	URI                   string         `mapstructure:"uri"`
+	Extract               map[string]any `mapstructure:"extract"`
+	GlobMaxTotalSize      int64          `mapstructure:"glob.max_total_size"`
+	GlobMaxObjectsMatched int            `mapstructure:"glob.max_objects_matched"`
+	GlobMaxObjectsListed  int64          `mapstructure:"glob.max_objects_listed"`
+	GlobPageSize          int            `mapstructure:"glob.page_size"`
+	url                   *globutil.URL
+	extractPolicy         *rillblob.ExtractPolicy
+}
+
+func parseSourceProperties(props map[string]any) (*sourceProperties, error) {
+	conf := &sourceProperties{}
+	err := mapstructure.WeakDecode(props, conf)
+	if err != nil {
+		return nil, err
+	}
+	if !doublestar.ValidatePattern(conf.Path) {
+		return nil, fmt.Errorf("glob pattern %s is invalid", conf.Path)
+	}
+	bucketURL, err := globutil.ParseBucketURL(conf.Path)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse path %q, %w", conf.Path, err)
+	}
+	if bucketURL.Scheme != "azure" {
+		return nil, fmt.Errorf("invalid scheme %q in path %q", bucketURL.Scheme, conf.Path)
+	}
+
+	conf.url = bucketURL
+	return conf, nil
+}
+
+// getClient returns a new azure blob client.
+func (c *Connection) getClient(ctx context.Context, conf *sourceProperties) (*container.Client, error) {
+	var accountName, accountKey, sasToken, connectionString string
+
+	if c.config.AllowHostAccess {
+		accountName = os.Getenv("AZURE_STORAGE_ACCOUNT")
+		accountKey = os.Getenv("AZURE_STORAGE_KEY")
+		sasToken = os.Getenv("AZURE_STORAGE_SAS_TOKEN")
+		connectionString = os.Getenv("AZURE_STORAGE_CONNECTION_STRING")
+	}
+
+	if c.config.Account != "" {
+		accountName = c.config.Account
+	}
+	if c.config.Key != "" {
+		accountKey = c.config.Key
+	}
+	if c.config.SASToken != "" {
+		sasToken = c.config.SASToken
+	}
+	if c.config.ConnectionString != "" {
+		connectionString = c.config.ConnectionString
+	}
+
+	if conf.AccountName != "" {
+		accountName = conf.AccountName
+	} else {
+		conf.AccountName = accountName
+	}
```

Comment:
This doesn't immediately appear necessary, but then I realized there's a very hard to see dependency here for propagating the correct account name to `createAnonymousClient`.

I would suggest refactoring the handling of the account name to make this simple and clear (even duplicating the logic would be better than this hidden dependency)